### PR TITLE
feat(Dynamic components): Added temporary registry type

### PR DESCRIPTION
**Pull request by m-suchorski**
Added a temporary registry, which is used in component registration process. Entires from temporary registry are moved to permanent one after a successful action. This approach eliminates the error of trying to attach component to a node that has already been initialized.

### CODE REVIEW

File: src/core/registry.js [hunk 1]
```diff
@@ -9,40 +9,64 @@ class Registry {
    */
   constructor() {
     this._registry = {};
+    this._tempRegistry = {};
```
Comment (xolir) [inline src/core/registry.js]: I would consider naming that registrationQueue to make it more clear what it does and how it differs from registry

File: src/core/registry.js [hunk 2]
```diff
@@ -9,40 +9,64 @@ class Registry {
...
   }
 
   /**
    * Retunrs all registry data
```
Comment (xolir) [inline src/core/registry.js]: There is typ here in comment, not directly connected to code that you changed

Comment (xolir) [reply src/core/registry.js]: Also, this should be updated to say that this returns both registry and temporary registry

File: src/core/registry.js [hunk 3]
```diff
@@ -9,40 +9,64 @@ class Registry {
...
   getRegisteredSelectors() {
     return Object
       .keys(this._registry);
   }
 
+  /**
+   * Returns an Array of temporary registry entires
+   * @returns {Array} registry entries
+   */
+  getNewlyRegisteredSelectors() {
```
Comment (xolir) [inline src/core/registry.js]: Consider renaming that aswell, see comment above

File: src/core/linker.js [hunk 4]
```diff
@@ -51,6 +51,7 @@ class Linker {
           const data = element.data();
           const Instance = this.registry.getComponent(selector);
           el.__strudel__ = new Instance({ element, data });
+          this.registry.setSelectorAsRegistered(selector);
```
Comment (xolir) [inline src/core/linker.js]: Should this be in this line? It will run for every element, not selector

File: src/core/registry.js [hunk 5]
```diff
@@ -9,40 +9,64 @@ class Registry {
...
+    this._tempRegistry = {};
   }
 
   /**
    * Retunrs all registry data
    * @returns {{}|*}
    */
   getData() {
-    return this._registry;
+    return Object.assign(this._registry, this._tempRegistry);
```
Comment (mateuszluczak) [inline src/core/registry.js]: IE11 doesn't like that

File: src/core/linker.js [hunk 6]
```diff
@@ -40,7 +40,7 @@ class Linker {
    * @param {DOMElement} container
    */
   link(container = document) {
-    this.registry.getRegisteredSelectors().forEach((selector) => {
+    this.registry.getNewlyRegisteredSelectors().forEach((selector) => {
```
Comment (mateuszluczak) [inline src/core/linker.js]: What if we add to DOM component that was already registered before? Will it not be initialised?

File: src/core/registry.js (hunk 5 above)
Comment (m-suchorski) [reply src/core/registry.js]: Oops, I must've forgot about that bi... beauty.

File: src/core/registry.js (hunk 1 above)
Comment (m-suchorski) [reply src/core/registry.js]: Done.

File: src/core/registry.js (hunk 2 above)
Comment (m-suchorski) [reply src/core/registry.js]: Done.

File: src/core/registry.js (hunk 3 above)
Comment (m-suchorski) [reply src/core/registry.js]: Done.

File: src/core/linker.js (hunk 4 above)
Comment (m-suchorski) [reply src/core/linker.js]: Done.

File: src/core/linker.js (hunk 6 above)
Comment (m-suchorski) [reply src/core/linker.js]: That has been changed to support both cases now, please take a look at the new approach if you don't mind.

File: src/core/registry.js (hunk 5 above)
Comment (m-suchorski) [reply src/core/registry.js]: I've created a workaround/polyfill solution, please take a look at it.

File: src/core/registry.js [hunk 7]
```diff
@@ -9,40 +9,73 @@ class Registry {
...
+    const registryArray = [this._registry, this._registrationQueue];
+
+    const mergedRegistry = registryArray.reduce((prev, curr) => {
+      Object.keys(curr).forEach((key) => {
+        prev[key] = curr[key];
+      });
+      return prev;
+    });
+
+    return mergedRegistry;
```
Comment (mateuszluczak) [inline src/core/registry.js]: Can we skip `registryArray` and `mergedRegistry` variables and return the reduce result?

File: src/core/linker.js [hunk 8]
```diff
@@ -55,6 +61,10 @@ class Linker {
           warn(`Trying to attach component to already initialized node, component with selector ${selector} will not be attached`);
         }
       });
+
+      if (isRootNode) {
```
Comment (mateuszluczak) [inline src/core/linker.js]: I don't understand this logic, can you elaborate why we need to distinguish rootNode from other nodes?

File: src/core/registry.js [hunk 9]
```diff
@@ -9,40 +9,73 @@ class Registry {
...
+    return Object
+      .keys(this._registrationQueue);
+  }
+
+  /**
+   * Moves selected registry entry from temporary to permanent
+   * @param {string} selector
+   */
+  setSelectorAsRegistered(selector) {
+    this._registry[selector] = this._registrationQueue[selector];
```
Comment (mateuszluczak) [inline src/core/registry.js]: Do we need to do that per selector or can we just push everything from queue to registry and just clear the queue when we are done?

File: test/e2e/specs/dynamic-cmp.js [hunk 10]
```diff
@@ -0,0 +1,11 @@
+module.exports = {
+  dynamicComponent(browser) {
```
Comment (mateuszluczak) [inline test/e2e/specs/dynamic-cmp.js]: I would call it "lazy registration"

File: src/core/linker.js (hunk 8 above)
Comment (xolir) [reply src/core/linker.js]: This is used to distinguish linker calls from component initalization when whole document is being traversed to find nodes and lazy HTML loading when only subtree is being crawled. That involves using registrationQueue and registry respectively.

Comment (mateuszluczak) [reply src/core/linker.js]: Nope, still don't get it 🤷‍♂

File: src/core/registry.js (hunk 7 above)
Comment (m-suchorski) [reply src/core/registry.js]: Fixed.

File: src/core/registry.js (hunk 9 above)
Comment (m-suchorski) [reply src/core/registry.js]: Fixed.

File: test/e2e/specs/dynamic-cmp.js (hunk 10 above)
Comment (m-suchorski) [reply test/e2e/specs/dynamic-cmp.js]: Done.

File: src/core/registry.js [hunk 11]
```diff
@@ -9,40 +9,71 @@ class Registry {
...
+    return Object
+      .keys(this._registrationQueue);
+  }
+
+  /**
+   * Moves selected registry entry from temporary to permanent
+   * @param {string} selector
+   */
+  setSelectorsAsRegistered() {
+    this._registry = this._registrationQueue;
```
Comment (mateuszluczak) [inline src/core/registry.js]: Don't you override what was already present here?
Shouldn't you merge two objects instead?

Comment (m-suchorski) [reply src/core/registry.js]: You're right, my bad. Gonna fix that rn.

Comment (m-suchorski) [reply src/core/registry.js]: I've created a helper function as merging is used in two places already, I hope it's okay.

File: src/core/linker.js (hunk 8 above)
Comment (mateuszluczak) [reply src/core/linker.js]: @xolir @m-suchorski 
So what do we do about it, can we discuss it irl?

Comment (m-suchorski) [reply src/core/linker.js]: Sure thing, I'll do my best in explaining this wicked logic

File: src/core/registry.js [hunk 12]
```diff
@@ -67,6 +68,15 @@ class Registry {
       warn(`Component registered under selector: ${selector} already exists.`, klass);
     } else {
       this._registrationQueue[selector] = klass;
+
+      if (!this._isRegisterScheduled) {
+        this._isRegisterScheduled = true;
+
+        setTimeout(() => {
```
Comment (xolir) [inline src/core/registry.js]: Can this be requestAnimationFrame or requestIdleCallback instead?

File: src/core/registry.js [hunk 13]
```diff
@@ -67,6 +68,15 @@ class Registry {
       warn(`Component registered under selector: ${selector} already exists.`, klass);
     } else {
       this._registrationQueue[selector] = klass;
+
+      if (!this._isRegisterScheduled) {
+        this._isRegisterScheduled = true;
```
Comment (xolir) [inline src/core/registry.js]: I would call this isRegistrationScheduled

Comment (m-suchorski) [reply src/core/registry.js]: Done.

File: src/core/registry.js (hunk 12 above)
Comment (m-suchorski) [reply src/core/registry.js]: requestIdleCallback isn't supported on IE11. I've cahnged it to requestAnimationFrame.

File: src/core/registry.js [hunk 14]
```diff
@@ -9,40 +10,74 @@ class Registry {
...
-     */
+   */
   registerComponent(selector, klass) {
-    if (this._registry[selector]) {
+    if (this._registry[selector] || this._registrationQueue[selector]) {
       warn(`Component registered under selector: ${selector} already exists.`, klass);
+    } else {
+      this._registrationQueue[selector] = klass;
+
+      if (!this._isRegistrationScheduled) {
```
Comment (mateuszluczak) [inline src/core/registry.js]: Don't we now have two registration cycles during initial page load?

Comment (m-suchorski) [reply src/core/registry.js]: Actually, yes. I'll try to fix that.

Comment (m-suchorski) [reply src/core/registry.js]: I think it's fixed now, please take a look.

File: src/core/init.js [hunk 15]
```diff
@@ -71,8 +71,6 @@ const onAutoTeardownCallback = (mutation) => {
 const init = () => {
   if (/comp|inter|loaded/.test(document.readyState)) {
     setTimeout(bootstrap, 0);
-  } else {
-    channel.on('DOMContentLoaded', bootstrap);
```
Comment (mateuszluczak) [inline src/core/init.js]: Why this become removed? Is there now a possibility when we initialise before DOM is ready? We should always wait for that

Comment (m-suchorski) [reply src/core/init.js]: I took a different approach this time by adding a check in the linker function, so it doesn't run linker event on document twice. Tell me what you think.

Comment (mateuszluczak) [reply src/core/init.js]: Let's meet IRL, it would be useful to create activity diagram - even from maintainers perspective

**Comment by m-suchorski** [comment]
I've tested examples in the way we've specified last week and the behaviour hasn't changed comparing to the current dev branch, so I guess we're good to go. I'd also like to test it in a real-world project, but that might take some additional time. Please tell me what you think @mateuszluczak

**Comment by mateuszluczak** [comment]
> I've tested examples in the way we've specified last week and the behaviour hasn't changed comparing to the current dev branch, so I guess we're good to go. I'd also like to test it in a real-world project, but that might take some additional time. Please tell me what you think @mateuszluczak

Ok, let's merge it and when we release 1.0.0-beta/rc we will check how it behaves

**Pull request merged**